src/panels/ConfigPanel/ConfirmationDialog/ConfirmationDialog.tsx

```diff
@@ -0,0 +1,203 @@
+    if (
+      canManageSecrets &&
+      getRequiredGrants(appName, config, secrets)?.length
+    ) {
+      setConfirmType(ConfigConfirmType.GRANT);
+    } else {
+      handleRemovePanelQueryParams();
+    }
+  }
+
+  if (appName && confirmType === DefaultConfirmType.SUBMIT) {
+    // Render the submit confirmation modal.
+    return (
+      <Portal>
+        <ConfirmationModal
+          // Prevent clicks inside this panel from closing the parent panel.
+          // This is handled in `checkCanClose`.
+          className="prevent-panel-close"
+          title={Label.SAVE_CONFIRM}
+          confirmExtra={
+            <p className="u-text--muted p-text--small u-align--left">
+              You can revert back to the applications default settings by
+              clicking the “Reset all values” button; or reset each edited field
+              by clicking “Use default”.
+            </p>
+          }
+          cancelButtonLabel={Label.SAVE_CONFIRM_CANCEL_BUTTON}
+          confirmButtonLabel={Label.SAVE_CONFIRM_CONFIRM_BUTTON}
+          confirmButtonAppearance="positive"
+          onConfirm={() => {
+            setConfirmType(null);
+            // Clear the form errors if there were any from a previous submit.
+            setInlineError(InlineErrors.FORM, null);
+            _submitToJuju().catch((error) => {
+              setInlineError(
+                InlineErrors.SUBMIT_TO_JUJU,
+                Label.SUBMIT_TO_JUJU_ERROR,
+              );
+              console.error(Label.SUBMIT_TO_JUJU_ERROR, error);
+            });
+          }}
+          close={() => setConfirmType(null)}
+        >
+          <ChangedKeyValues appName={appName} config={config} />
+        </ConfirmationModal>
+      </Portal>
+    );
+  } else if (appName && confirmType === ConfigConfirmType.GRANT) {
+    // Render the grant confirmation modal.
+    const requiredGrants = getRequiredGrants(appName, config, secrets);
+    return (
+      <Portal>
+        <ConfirmationModal
+          // Prevent clicks inside this panel from closing the parent panel.
+          // This is handled in `checkCanClose`.
+          className="prevent-panel-close"
+          title={Label.GRANT_CONFIRM}
+          cancelButtonLabel={Label.GRANT_CANCEL_BUTTON}
+          confirmButtonLabel={Label.GRANT_CONFIRM_BUTTON}
+          confirmButtonAppearance="positive"
+          onConfirm={() => {
+            setConfirmType(null);
+            // Clear the form errors if there were any from a previous submit.
+            setInlineError(InlineErrors.FORM, null);
+            if (!appName || !requiredGrants) {
+              // It is not possible to get to this point if these
+              // variables aren't set.
+              return;
+            }
+            void (async () => {
+              try {
+                for (const secretURI of requiredGrants) {
+                  await grantSecret(secretURI, [appName]);
+                }
+                setConfirmType(null);
```

Comment:
This call might be redundant as well, as we set the confirm type to `null` on line 137.